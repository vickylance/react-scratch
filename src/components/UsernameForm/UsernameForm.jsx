import React, { useState } from 'react';
import PropTypes from 'prop-types';

const UsernameForm = ({ updateUsername }) => {
  const [status, setStatus] = useState('idle');
  const [error, setError] = useState(null);
  const usernameId = 'username';

  const handleSubmit = async (e) => {
    e.preventDefault();
    const newUsername = e.target.elements.username.value;
    setStatus('pending');
    try {
      await updateUsername(newUsername);
      setStatus('fulfilled');
    } catch (err) {
      setStatus('rejected');
      setError(err);
    }
  };

  return (
    <form onSubmit={handleSubmit}>
      <div>
        <label htmlFor={usernameId}>
          Username
          <input type="text" id={usernameId} />
        </label>
      </div>
      <button type="submit">Submit</button>
      <span>{status === 'pending' ? 'Saving...' : null}</span>
      <span>{status === 'rejected' ? error.message : null}</span>
    </form>
  );
};

UsernameForm.propTypes = {
  updateUsername: PropTypes.func.isRequired,
};
UsernameForm.defaultProps = {};

export default UsernameForm;
