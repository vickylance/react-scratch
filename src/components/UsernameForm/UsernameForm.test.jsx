import React from 'react';
import user from '@testing-library/user-event';
import { render, screen, waitForElementToBeRemoved } from '@testing-library/react';
import UsernameForm from './UsernameForm';

test('renders properly', () => {
  const handleUpdateUsername = jest.fn(() => Promise.resolve());
  const { container } = render(<UsernameForm updateUsername={handleUpdateUsername} />);
  expect(container).toMatchSnapshot();
});

test('calls updateUsername with the new username', async () => {
  const handleUpdateUsername = jest.fn(() => Promise.resolve());
  const fakeUsername = 'sonicTheHedgehog';

  render(<UsernameForm updateUsername={handleUpdateUsername} />);
  const usernameInput = screen.getByLabelText(/username/i);
  user.type(usernameInput, fakeUsername);
  user.click(screen.getByText(/submit/i));

  expect(handleUpdateUsername).toBeCalledWith(fakeUsername);
  await waitForElementToBeRemoved(() => screen.queryByText(/Saving.../i));
});
