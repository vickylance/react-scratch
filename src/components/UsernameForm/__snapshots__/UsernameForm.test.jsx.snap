// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders properly 1`] = `
<div>
  <form>
    <div>
      <label
        for="username"
      >
        Username
        <input
          id="username"
          type="text"
        />
      </label>
    </div>
    <button
      type="submit"
    >
      Submit
    </button>
    <span />
    <span />
  </form>
</div>
`;
